src/Dependencies/Collections/RoslynImmutableInterlocked.cs

```diff
@@ -0,0 +1,369 @@
+        /// the same because the last invocation of <paramref name="transformer"/> returned the existing value.
+        /// </returns>
+        public static bool Update<TKey, TValue>(ref ImmutableSegmentedDictionary<TKey, TValue> location, Func<ImmutableSegmentedDictionary<TKey, TValue>, ImmutableSegmentedDictionary<TKey, TValue>> transformer)
+            where TKey : notnull
+        {
+            if (transformer is null)
+                throw new ArgumentNullException(nameof(transformer));
+
+            bool successful;
+            var oldValue = ImmutableSegmentedDictionary<TKey, TValue>.PrivateInterlocked.VolatileRead(ref location);
+            do
+            {
+                var newValue = transformer(oldValue);
+                if (oldValue == newValue)
+                {
+                    // No change was actually required.
+                    return false;
+                }
+
+                var interlockedResult = InterlockedCompareExchange(ref location, newValue, oldValue);
+                successful = oldValue == interlockedResult;
+                oldValue = interlockedResult; // we already have a volatile read that we can reuse for the next loop
+            }
+            while (!successful);
+
+            return true;
+        }
+
+        /// <summary>
+        /// Mutates a value in-place with optimistic locking transaction semantics via a specified transformation
+        /// function. The transformation is retried as many times as necessary to win the optimistic locking race.
+        /// </summary>
+        /// <typeparam name="TKey">The type of key stored by the dictionary.</typeparam>
+        /// <typeparam name="TValue">The type of value stored by the dictionary.</typeparam>
+        /// <typeparam name="TArg">The type of argument passed to the <paramref name="transformer"/>.</typeparam>
+        /// <param name="location">
+        /// The variable or field to be changed, which may be accessed by multiple threads.
+        /// </param>
+        /// <param name="transformer">
+        /// A function that mutates the value. This function should be side-effect free, as it may run multiple times
+        /// when races occur with other threads.</param>
+        /// <param name="transformerArgument">The argument to pass to <paramref name="transformer"/>.</param>
+        /// <returns>
+        /// <see langword="true"/> if the location's value is changed by applying the result of the
+        /// <paramref name="transformer"/> function; otherwise, <see langword="false"/> if the location's value remained
+        /// the same because the last invocation of <paramref name="transformer"/> returned the existing value.
+        /// </returns>
+        public static bool Update<TKey, TValue, TArg>(ref ImmutableSegmentedDictionary<TKey, TValue> location, Func<ImmutableSegmentedDictionary<TKey, TValue>, TArg, ImmutableSegmentedDictionary<TKey, TValue>> transformer, TArg transformerArgument)
```

Comment:
>TArg transformerArgument [](start = 223, length = 24)

A long-standing pet-peeve of mine with our generic APIs is that we put the argument last, which means that nothing can be inferred about the lambda argument unless you write the 3rd parameter, then go back and write the transformation lambda. Consider making this the second parameter to make typing more ergonomic.

Reply:
➡️ This class intentionally follows the pattern from `ImmutableInterlocked`. More specifically, code which currently uses `ImmutableInterlocked`+`ImmutableDictionary<TKey, TValue>` can switch to `ImmutableSegmentedDictionary<TKey, TValue>` by simply adding the `Roslyn` prefix. (Drop-in replacement is a primary design goal.)

Reply:
I agree that the signature is not ideal, but also that there is a precedent that we should follow. It'd worse if different APIs used a different order.

---
In reply to: [561169365](https://github.com/dotnet/roslyn/pull/50189#discussion_r561169365) [](ancestors = 561169365)